File: facebook-scraper/scrape_fb.js
#!/usr/bin/env node
/**
 * Facebook Comment Scraper
 *
 * Setup (run once):
 *   npm install
 *   npx playwright install chromium
 *   node scrape_fb.js --setup
 *
 * Scrape a post:
 *   node scrape_fb.js "https://www.facebook.com/..."
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const SCRIPT_DIR = __dirname;
const SESSION_FILE = path.join(SCRIPT_DIR, 'fb_session.json');

function waitForEnter(prompt) {
  return new Promise(resolve => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    rl.question(prompt, () => { rl.close(); resolve(); });
  });
}

function timestamp() {
  const d = new Date();
  const pad = n => String(n).padStart(2, '0');
  return `${d.getFullYear()}-${pad(d.getMonth()+1)}-${pad(d.getDate())}_${pad(d.getHours())}${pad(d.getMinutes())}${pad(d.getSeconds())}`;
}

function escapeCSV(val) {
  if (val === null || val === undefined) return '';
  const str = String(val);
  if (str.includes('"') || str.includes(',') || str.includes('\n')) {
    return '"' + str.replace(/"/g, '""') + '"';
  }
  return str;
}

async function setupSession() {
  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext();
  const page = await context.newPage();

  await page.goto('https://www.facebook.com/');

  console.log('\nA browser window has opened.');
  console.log('Log into Facebook in the browser, then come back here and press Enter.');
  await waitForEnter('Press Enter when you are logged in... ');

  const cookies = await context.cookies();
  fs.writeFileSync(SESSION_FILE, JSON.stringify(cookies, null, 2));
  await browser.close();

  console.log('\nSession saved to fb_session.json');
  console.log('You can now run:  node scrape_fb.js "<post-url>"');
}

async function loadAllComments(page) {
  let clicks = 0;
  for (let i = 0; i < 100; i++) {
    try {
      const btn = page.locator('text=/more comments/i').first();
      if (await btn.isVisible({ timeout: 3000 })) {
        await btn.scrollIntoViewIfNeeded();
        await btn.click();
        await page.waitForTimeout(2000);
        clicks++;
        process.stdout.write(`  Loading comments... (${clicks} batches)\r`);
      } else {
        break;
      }
    } catch {
      break;
    }
  }
  if (clicks) console.log(`  Loading comments... (${clicks} batches total)    `);
}

async function expandAllReplies(page) {
  for (let i = 0; i < 50; i++) {
    const buttons = await page.locator('text=/\\d+ repl/i').all();
    let expandedAny = false;
    for (const btn of buttons) {
      try {
        if (await btn.isVisible()) {
          await btn.scrollIntoViewIfNeeded();
          await btn.click();
          await page.waitForTimeout(1000);
          expandedAny = true;
        }
      } catch { /* skip */ }
    }
    if (!expandedAny) break;
  }
}

async function extractComments(page, postUrl) {
  const data = await page.evaluate(() => {
    const out = [];
    for (const el of document.querySelectorAll('[data-commentid]')) {
      const authorEl = el.querySelector('a[role="link"]');
      const author = authorEl ? authorEl.innerText.trim() : '';

      let commentText = '';
      for (const div of el.querySelectorAll('div[dir="auto"]')) {
        const t = div.innerText.trim();
        if (t && t !== author) { commentText = t; break; }
      }

      let ts = '';
      for (const a of el.querySelectorAll('a[aria-label]')) {
        const lbl = a.getAttribute('aria-label') || '';
        if (lbl && /\d/.test(lbl)) { ts = lbl; break; }
      }
      if (!ts) {
        const abbr = el.querySelector('abbr[title]');
        if (abbr) ts = abbr.getAttribute('title') || '';
      }

      let likes = '0';
      for (const span of el.querySelectorAll('[aria-label]')) {
        const lbl = span.getAttribute('aria-label') || '';
        if (/reaction|like/i.test(lbl)) {
          const m = lbl.match(/\d+/);
          if (m) { likes = m[0]; break; }
        }
      }

      const parentComment = el.parentElement
        ? el.parentElement.closest('[data-commentid]')
        : null;
      const isReply = !!(parentComment && parentComment !== el);
      let replyTo = '';
      if (isReply) {
        const pAuthor = parentComment.querySelector('a[role="link"]');
        replyTo = pAuthor ? pAuthor.innerText.trim() : '';
      }

      if (author || commentText) {
        out.push({ author, commentText, ts, likes, isReply, replyTo });
      }
    }
    return out;
  });

  return data.map(c => ({ ...c, postUrl }));
}

async function scrape(url) {
  if (!fs.existsSync(SESSION_FILE)) {
    console.log('No session found. Run setup first:\n  node scrape_fb.js --setup');
    process.exit(1);
  }

  const cookies = JSON.parse(fs.readFileSync(SESSION_FILE, 'utf8'));

  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext();
  await context.addCookies(cookies);
  const page = await context.newPage();

  console.log('Opening post...');
  await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
  await page.waitForTimeout(3000);

  if (page.url().includes('login') && page.url().includes('facebook.com')) {
    console.log('\nSession expired — run --setup again to log in.');
    await browser.close();
    process.exit(1);
  }

  console.log('Loading all comments (may take a moment)...');
  await loadAllComments(page);

  console.log('Expanding replies...');
  await expandAllReplies(page);

  console.log('Extracting data...');
  const comments = await extractComments(page, url);
  await browser.close();

  if (!comments.length) {
    console.log('\nNo comments found.');
    console.log('Possible reasons:');
    console.log('  - The post is private or requires login to see comments');
    console.log('  - Facebook changed its layout (update selectors in extractComments)');
    return;
  }

  const outFile = path.join(SCRIPT_DIR, `comments_${timestamp()}.csv`);
  const header = 'author,comment_text,timestamp,likes,is_reply,reply_to,post_url';
  const rows = comments.map(c => [
    escapeCSV(c.author),
    escapeCSV(c.commentText),
    escapeCSV(c.ts),
    escapeCSV(c.likes),
    escapeCSV(c.isReply),
    escapeCSV(c.replyTo),
    escapeCSV(c.postUrl),
  ].join(','));

  fs.writeFileSync(outFile, [header, ...rows].join('\n'), 'utf8');
  console.log(`\nSaved ${comments.length} comments to ${path.basename(outFile)}`);
}

const args = process.argv.slice(2);
if (!args.length || args[0] === '--help') {
  console.log('Usage:');
  console.log('  node scrape_fb.js --setup        # Set up Facebook login (run once)');
  console.log('  node scrape_fb.js "<url>"        # Scrape comments from a post');
  process.exit(args.length ? 0 : 1);
} else if (args[0] === '--setup') {
  setupSession().catch(e => { console.error(e); process.exit(1); });
} else {
  scrape(args[0]).catch(e => { console.error(e); process.exit(1); });
}

File: facebook-scraper/scrape_fb.py
#!/usr/bin/env python3
"""
Facebook Comment Scraper

Setup (run once):
  pip install playwright
  playwright install chromium
  python scrape_fb.py --setup

Scrape a post:
  python scrape_fb.py "https://www.facebook.com/..."
"""

import asyncio
import csv
import json
import sys
from datetime import datetime
from pathlib import Path

from playwright.async_api import async_playwright

SCRIPT_DIR = Path(__file__).parent
SESSION_FILE = SCRIPT_DIR / "fb_session.json"


async def setup_session():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context()
        page = await context.new_page()

        await page.goto("https://www.facebook.com/")

        print("\nA browser window has opened.")
        print("Log into Facebook in the browser, then come back here and press Enter.")
        input("Press Enter when you are logged in...")

        cookies = await context.cookies()
        SESSION_FILE.write_text(json.dumps(cookies, indent=2))
        await browser.close()

        print(f"\nSession saved to fb_session.json")
        print('You can now run:  python scrape_fb.py "<post-url>"')


async def load_all_comments(page):
    """Click 'Load more comments' repeatedly until no more appear."""
    clicks = 0
    for _ in range(100):
        try:
            btn = page.locator("text=/more comments/i").first
            if await btn.is_visible(timeout=3000):
                await btn.scroll_into_view_if_needed()
                await btn.click()
                await page.wait_for_timeout(2000)
                clicks += 1
                print(f"  Loading comments... ({clicks} batches)", end="\r")
            else:
                break
        except Exception:
            break
    if clicks:
        print(f"  Loading comments... ({clicks} batches total)    ")


async def expand_all_replies(page):
    """Click every 'View X replies' link until none remain."""
    for _ in range(50):
        try:
            buttons = await page.locator("text=/\\d+ repl/i").all()
            expanded_any = False
            for btn in buttons:
                try:
                    if await btn.is_visible():
                        await btn.scroll_into_view_if_needed()
                        await btn.click()
                        await page.wait_for_timeout(1000)
                        expanded_any = True
                except Exception:
                    continue
            if not expanded_any:
                break
        except Exception:
            break


async def extract_comments(page, post_url):
    """
    Run JavaScript inside the loaded page to pull comment data out of the DOM.
    Uses data-commentid attributes which are more stable than Facebook's generated class names.
    """
    data = await page.evaluate("""() => {
        const out = [];

        for (const el of document.querySelectorAll('[data-commentid]')) {
            // Author — first link with role=link inside this comment element
            const authorEl = el.querySelector('a[role="link"]');
            const author = authorEl ? authorEl.innerText.trim() : '';

            // Comment text — first non-empty div[dir=auto] that isn't just the author's name
            let commentText = '';
            for (const div of el.querySelectorAll('div[dir="auto"]')) {
                const t = div.innerText.trim();
                if (t && t !== author) {
                    commentText = t;
                    break;
                }
            }

            // Timestamp — look for an aria-label on a link that contains a digit
            let timestamp = '';
            for (const a of el.querySelectorAll('a[aria-label]')) {
                const lbl = a.getAttribute('aria-label') || '';
                if (lbl && /\\d/.test(lbl)) {
                    timestamp = lbl;
                    break;
                }
            }
            if (!timestamp) {
                const abbr = el.querySelector('abbr[title]');
                if (abbr) timestamp = abbr.getAttribute('title') || '';
            }

            // Like / reaction count
            let likes = '0';
            for (const span of el.querySelectorAll('[aria-label]')) {
                const lbl = span.getAttribute('aria-label') || '';
                if (/reaction|like/i.test(lbl)) {
                    const m = lbl.match(/\\d+/);
                    if (m) { likes = m[0]; break; }
                }
            }

            // Is this a reply? Replies are nested inside another [data-commentid] element
            const parentComment = el.parentElement
                ? el.parentElement.closest('[data-commentid]')
                : null;
            const isReply = !!(parentComment && parentComment !== el);
            let replyTo = '';
            if (isReply) {
                const pAuthor = parentComment.querySelector('a[role="link"]');
                replyTo = pAuthor ? pAuthor.innerText.trim() : '';
            }

            if (author || commentText) {
                out.push({ author, commentText, timestamp, likes, isReply, replyTo });
            }
        }

        return out;
    }""")

    for item in data:
        item['post_url'] = post_url

    return data


async def scrape(url):
    if not SESSION_FILE.exists():
        print("No session found. Run setup first:")
        print("  python scrape_fb.py --setup")
        sys.exit(1)

    cookies = json.loads(SESSION_FILE.read_text())

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context()
        await context.add_cookies(cookies)
        page = await context.new_page()

        print("Opening post...")
        await page.goto(url, wait_until="domcontentloaded", timeout=30000)
        await page.wait_for_timeout(3000)

        if "login" in page.url and "facebook.com" in page.url:
            print("\nSession expired — run --setup again to log in.")
            await browser.close()
            sys.exit(1)

        print("Loading all comments (may take a moment)...")
        await load_all_comments(page)

        print("Expanding replies...")
        await expand_all_replies(page)

        print("Extracting data...")
        comments = await extract_comments(page, url)
        await browser.close()

    if not comments:
        print("\nNo comments found.")
        print("Possible reasons:")
        print("  - The post is private or requires a friend connection to view comments")
        print("  - Facebook has changed its page layout (update the selectors in extract_comments)")
        return

    ts = datetime.now().strftime("%Y-%m-%d_%H%M%S")
    out_file = SCRIPT_DIR / f"comments_{ts}.csv"

    with open(out_file, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=[
            "author", "comment_text", "timestamp", "likes",
            "is_reply", "reply_to", "post_url"
        ])
        writer.writeheader()
        for c in comments:
            writer.writerow({
                "author": c["author"],
                "comment_text": c["commentText"],
                "timestamp": c["timestamp"],
                "likes": c["likes"],
                "is_reply": c["isReply"],
                "reply_to": c["replyTo"],
                "post_url": c["post_url"],
            })

    print(f"\nSaved {len(comments)} comments to {out_file.name}")


def main():
    if len(sys.argv) < 2 or sys.argv[1] in ("-h", "--help"):
        print(__doc__)
        sys.exit(0 if "--help" in sys.argv else 1)

    if sys.argv[1] == "--setup":
        asyncio.run(setup_session())
    else:
        asyncio.run(scrape(sys.argv[1]))


if __name__ == "__main__":
    main()
